### PR TITLE
Cleanup resolve

### PR DESCRIPTION
**Pull request by jseyfried**
This is a collection of small refactorings and improvements in `resolve`, most of which could stand alone.
r? @nikomatsakis

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @petrochenkov

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1440,40 +1445,66 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
+    ///
+    /// A block's items are above its local variables in the scope hierarchy, regardless of where
+    /// the items are defined in the block. For example,
+    /// ```rust
+    /// fn f() {
+    ///    g(); // Since there are no local variables in scope yet, this resolves to the item.
+    ///    let g = || {};
+    ///    fn g() {}
+    ///    g(); // This resolves to the local variable `g` since it shadows the item.
+    /// }
```
Comment (nikomatsakis) [inline src/librustc_resolve/lib.rs]: love to see more comments! :)

**Pull request merged**